### PR TITLE
Remove Send bounds from scopes.

### PR DESCRIPTION
**Pull request by Zoxc**

### CODE REVIEW

File: rayon-core/src/scope/mod.rs [hunk 1]
```diff
@@ -536,12 +530,14 @@ impl<'scope> ScopeBase<'scope> {
...
     ///
     /// Unsafe because it must be executed on a worker thread.
-    unsafe fn complete<FUNC, R>(&self, owner_thread: &WorkerThread, func: FUNC) -> R
+    unsafe fn complete<FUNC, R>(&self, func: FUNC) -> R
     where
         FUNC: FnOnce() -> R,
     {
         let result = self.execute_job_closure(func);
-        self.steal_till_jobs_complete(owner_thread);
+        in_worker(|owner_thread, _| {
```
Comment (cuviper) [inline rayon-core/src/scope/mod.rs:538]: It seems unfortunate to inject yet another job just to wait for completion. Maybe we could add another latch type that acts like a combination of `CountLatch` and `LockLatch`, raising the `Condvar` when the count is zero. Then `job_completed_latch` would need to be an enum of either `CountLatch` or this new thing, depending on what context it's called from.

**Review by cuviper — COMMENTED** [review]
We should probably reconsider this upstream -- see rayon-rs/rayon#562. Can you also add some description motivating this change in your case?

File: rayon-core/src/scope/mod.rs (hunk 1 above)
Comment (Zoxc) [reply rayon-core/src/scope/mod.rs:538]: This doesn't matter for rustc, since it always runs in a worker.

**Comment by Zoxc** [comment]
I want to run some code which refers to `!Send` types in parallel with some other code which do use `Send` types. Specifically name resolution (with `!Send`) and gated feature checking in rustc.

**Comment by nikomatsakis** [comment]
Closing this pull request as Zoxc is stepping back from compiler development; see rust-lang/team#316.

**Pull request closed without merging**